README.md
Wifitap
=======

	#########################################
	#
	# wifitap.py --- WiFi injection tool through tun/tap device
	# Cedric Blancher <sid@rstack.org>
	#
	# http://sid.rstack.org/index.php/Wifitap (french)
	# http://sid.rstack.org/index.php/Wifitap_EN (english)
	#
	#########################################

This program is a proof of concept tool allowing WiFi communications using
traffic injection.

You'll need:

	. Python >= 2.2
	. Philippe Biondi's Scapy
	. Injection ready wireless adapter

It's been tested on GNU/Linux using Atheros chipset based adapter with patched
Madwifi driver and Intersil Prism GT Full MACchipset with Prism54 driver. It
should as well work with Prism2/2.5/3 chipset hostap driver or wlan-ng driver,
Ralink rt2500/2750 chipset using rt2500 driver and Realtek RTL8180 chipset
using rtl8180-sa2400 driver.

I didn't take time to test Prism2/2.5/3 support and don't have Ralink or Realtek
based hardware for testing. By the way, I would be glad to have feedback for
Wifitap attempts with thoses chipsets.

Drivers patches are written by Christophe Devine and updated by Aircrack-ng
people. For details about drivers patch and installation, see PATCHING file.

To get wifitap work on other Unix operating systems than GNU/Linux, you have to
install pcap or dnet wrappers for Python so Scapy can work (see
http://www.secdev.org/projects/scapy/portability.html). Then, and it's the most
important part, you have to find a wireless adapter driver that supports raw
wireless traffic injection if any.

   * [Setup Guide](#setup-guide)
      * [Getting Wifitap ;)](#getting-wifitap-)
      * [Installing Dependencies](#installing-dependencies)
   * [Usage Guide](#usage-guide)
      * [Launching Wifitap](#launching-wifitap)
      * [Wifitap Command Line Arguments](#wifitap-command-line-arguments)

Setup Guide
===========

Getting Wifitap ;)
------------------

The latest Wifitap can be obtained from this Git repo.

Older versions of Wifitap are available at the following locations:

- [http://sid.rstack.org/index.php/Wifitap (french)](http://sid.rstack.org/index.php/Wifitap)
- [http://sid.rstack.org/index.php/Wifitap\_EN (english)](http://sid.rstack.org/index.php/Wifitap_EN)



Installing Dependencies
-----------------------

Python dependencies are enumerated in `pip.req`, can be installed using the pip as shown below.

	pip install -r pip.req

Wifitap's most important dependency is Scapy, which you should check out because it's totally awesome.

 - [http://www.secdev.org/projects/scapy/](http://www.secdev.org/projects/scapy/)


Usage Guide
===========

Launching Wifitap
-----------------

Before you start, make sure your wireless interface is in monitor mode. In the commands shown below, substitute "wlan0" for the name of your wireless interface.
	
	ifconfig wlan0 down
	iw dev wlan0 set monitor none
	ifconfig wlan0 up promisc

To launch Wifitap with basic options, use the following command:

	./wifitap.py -b <bssid>

This will create a wj0 interface. Next, you'll need to configure this new interface using syntax shown below. Note that you can optionally specify the MAC in the following command.

	ifconfig wj0 [hw ether <MAC>] 192.168.1.1 [mtu <MTU>]

You'll now be able to use your newly created wj0 interface as if it were normal network interface, according to your system routing table. :)


Wifitap Command Line Arguments
------------------------------

	Usage : wifitap -b <BSSID> [-o <iface>] [-i <iface> [-s <SMAC>]
			[-w <WEP key> [-k <key id>]] [-d [-v]] [-h]

	-b	Specifies BSSID in ususal 6 hex digits MAC address format:
			. 00:01:02:03:04:05

	-o	Specifies output WiFi interface for frames injection

	-i	Specifies input WiFi interface for frames sniffing

	-s	Specifies source MAC address
			. 00:01:02:03:04:05

	-w	Activates WEP encryption/decryption with specified WEP key
		Key can be given using following formats:
			. 0102030405 or 0102030405060708090a0b0c0d
			. 01:02:03:04:05 or
			  01:02:03:04:05:06:07:08:09:0a:0b:0c:0d
			. 0102-0304-05 or 0102-0304-0506-0708-090a-0b0c-0d

	-k	Specifies WEP key id, from 0 to 3

	-d	Activates debugging

	-v	Increases debugging verbosity

	-h	Help screen

	#########################################
	#
	# Copyright (C) 2005 Cedric Blancher <sid@rstack.org>
	#
	# This program is free software; you can redistribute it and/or modify it
	# under the terms of the GNU General Public License version 2 as
	# published by the Free Software Foundation; version 2.
	#
	# This program is distributed in the hope that it will be useful, but
	# WITHOUT ANY WARRANTY; without even the implied warranty of
	# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
	# General Public License for more details.
	#
	#########################################


Changelog
0.1   - Initial release

0.2   - BPF filter added on listening socket
	WEP support

0.3   - Code cleanup
	Arguments input and parsing
	Psyco optimizer support
	Public release

0.3.1 - Typo fixed

0.3.2 - Versioning changed
	Hostap and wlan-ng patches
	More documentation

0.3.3 - Updated patches from Christophe Devine for hostap and madwifi
	New patches from Christophe Devine for rt2500 and rtl8180
	Updated README

0.3.4 - Updated patches from Christophe Devine
	Updated documentation: README, PATCHING
	Added links to drivers and patches on http://100h.org/
	Added a sample ping answering machine (wifians.py)

0.3.5 - Updated documentation: README
	Renamed wifians.py into wifiping.py
	Added a sample DNS answering machine (wifidns.py)

0.3.6 - Code cleanup
	Awful bugs corrected in wifitap.py and wifidns.py
	Moved to Scapy v1.0.0.30

0.3.7 - Adjusted frame size for tuntap (1526) and 802.11 (2346)

0.4.0 - Updated documentation: README, PATCHING
	Prism Headers handling removed as latest libpcap handles them
	Added patch to enable Prism Headers if needed
	Moved to Scapy v1.0.4.74
	Added Scapy warnings support
	Code cleanup, optimizations and bugfixing (thx to P. Biondi)
	Added source MAC address setting support
	Added a sample ARP answering machine (wifiarp.py)

0.4.1 - Updated to use modern scapy and work with Linux radiotap headers
	Oliver Lavery, Gotham Digital Science (olavery@gdssecurity.com)

0.4.2 - Gabriel Ryan, Gotham Digital Science (gryan@gdssecurity.com), March 28th 2017
	Fixed indentation caused by mixing tabs and spaces. Project now uses spaces only.
	Fixed import statements to work with modern versions of Scapy.
	Removed unused imports.
	Psyco is no longer maintained and does not support Python 2.7. Removed from project in favor of PyPy.
	Code cleanup and optimizations.
	Project now uses the newer argparse module instead of getopt to parse command line arguments.
	Input validation is now done through argparse as much as possible.
	Added utils.py file containing commonly used utility functions.
	Added include guard to all files.
	Added pip requirements file.
	Updated README.


utils.py
import re
import argparse

def parse_wep_key(wepkey, keyid):

    print wepkey

    # Match and parse WEP key
    tmp_key = ""
    if re.match('^([0-9a-fA-F]{2}){5}$', wepkey) or re.match ('^([0-9a-fA-F]{2}){13}$', wepkey):
        tmp_key = wepkey
    elif re.match('^([0-9a-fA-F]{2}[:]){4}[0-9a-fA-F]{2}$', wepkey) or re.match('^([0-9a-fA-F]{2}[:]){12}[0-9a-fA-F]{2}$', wepkey):
        tmp_key = re.sub(':', '', wepkey)
    elif re.match ('^([0-9a-fA-F]{4}[-]){2}[0-9a-fA-F]{2}$', wepkey) or re.match ('^([0-9a-fA-F]{4}[-]){6}[0-9a-fA-F]{2}$', wepkey):
        tmp_key = re.sub('-', '', wepkey)
    else:
        return None
    g = lambda x: chr(int(tmp_key[::2][x],16)*16+int(tmp_key[1::2][x],16))

    conf_wepkey = ''
    for i in range(len(tmp_key)/2):
        conf_wepkey += g(i)

    return conf_wepkey

def sanitize_bssid(bssid):

    # Match and parse BSSID
    if re.match('^([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}$', bssid):
        return bssid.lower()

    msg = 'Invalid bssid.'
    raise argparse.ArgumentTypeError(msg)

def sanitize_smac(smac):

    if type(smac) == str and re.match('^([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}$', smac):
        return smac.lower()

    msg = 'Invalid SMAC.'
    raise argparse.ArgumentTypeError(msg)

def sanitize_ipdns(ipdns):

    if re.match('^(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])\.(\d|[1-9]\d|1\d\d|2[0-4]\d|25[0-5])$', ipdns):
        return ipdns

    msg = 'Error: Wrong IP address'
    raise argparse.ArgumentTypeError(msg)


wifiarp.py
#!/usr/bin/env python

########################################
#
# wifiping.py --- WiFi injection based answering tool based on Wifitap
#
# Copyright (C) 2005 Cedric Blancher <sid@rstack.org>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation; version 2.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
#########################################

import os
import sys
import argparse
import struct
import logging
import utils

from socket import *
from fcntl  import ioctl
from select import select

logging.getLogger("scapy").setLevel(1)
from scapy.all  import Raw,Ether,PrismHeader,Dot11,Dot11WEP,LLC,SNAP,sendp,conf,ARP

def setup():

    global conf

    parser = argparse.ArgumentParser()
    parser.add_argument('-b',
                    dest='bssid',
                    type=utils.sanitize_bssid,
                    required=True,
                    help='Specify BSSID for injection')

    parser.add_argument('-i',
                    dest='in_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for injection (default: wlan0)')

    parser.add_argument('-o',
                    dest='out_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for injection (default: wlan0)')

    parser.add_argument('-s',
                    dest='smac',
                    type=utils.sanitize_smac,
                    help='Specify source MAC address for injected frames.')

    parser.add_argument('-w',
                    dest='wepkey',
                    type=str,
                    help='WEP mode and key')

    parser.add_argument('-k',
                    dest='keyid',
                    type=int,
                    choices=xrange(0,4),
                    default=0,
                    help='WEP key id (default: 0)')

    parser.add_argument('-d',
                    dest='debug',
                    action='store_true',
                    help='Activate debug mode.')

    parser.add_argument('-v',
                    dest='verb',
                    action='store_true',
                    help='Use verbose debugging.')

    args = parser.parse_args()

    # validate wepkey if WEP in use
    if args.wepkey is not None:
        wepkey = utils.parse_wep_key(args.wepkey, args.keyid)
        if wepkey is None:
            parser.print_usage()
            print
            print '[!] Invalid WEP key'
            sys.exit()
    
    conf.iface = args.out_iface
    conf.wepkey = args.wepkey

    return args

def print_options(options):

    print "out_iface:  %s" % options.out_iface
    print "bssid:      %s" % options.bssid
    if options.smac is not None:
    
        print "smac:       %s" % options.smac
    
    if options.wepkey is not None:

        print "WEP key:    %s (%dbits)" % (options.wepkey, len(options.wepkey)*4)

    print 'DEBUG mode:', options.debug

if __name__ == '__main__':

    options = setup()

    # Here we put a BPF filter so only 802.11 Data/to-DS frames are captured
    s = conf.L2listen(iface = options.in_iface,
                filter = "link[0]&0xc == 8 and link[1]&0xf == 1")
    
    # Speed optimization si Scapy does not have to parse payloads
    ARP.payload_guess=[]
    
    try:
        while True:
            dot11_frame = s.recv(2346)
    
            # WEP handling is automagicly done by Scapy if conf.wepkey is set
            # Nothing to do to decrypt (although not yet tested)
            # WEP frames have Dot11WEP layer, others don't
            if options.debug and options.verb:
                if dot11_frame.haslayer(Dot11WEP): # WEP frame
                    os.write(1,"Received WEP from %s\n" % options.in_iface)
                else: # Cleartext frame
                    os.write(1,"Received from %s\n" % options.in_iface)
                    os.write(1,"%s\n" % dot11_frame.summary())
            
            if dot11_frame.getlayer(Dot11).addr1 != options.bssid:
                continue
            
            # Identifying ARP Requests
            if dot11_frame.haslayer(ARP) and dot11_frame.getlayer(ARP).op == 1:
    
                if options.debug:
    
                    os.write(1,"Received ARP Request on %s\n" % options.in_iface)
    
                    if optios.verb:
                        os.write(1,"%s\n" % dot11_frame.summary())
            
                # Building ARP Reply answer for injection
                dot11_answer = RadioTap()/Dot11(
                type = "Data",
                FCfield = "from-DS",
                addr1 = dot11_frame.getlayer(Dot11).addr2,
                addr2 = options.bssid)
                dot11_answer.addr3 = options.smac
    
                if options.wepkey is not None:
                    dot11_answer.FCfield |= 0x40
                    dot11_answer /= Dot11WEP(
                        iv = "111",
                        keyid = options.keyid)
    
                dot11_answer /= LLC(ctrl=3)/SNAP()/ARP(
            	                                    op="is-at",
            	                                    hwsrc=options.smac,
            	                                    psrc=dot11_frame.getlayer(ARP).pdst,
            	                                    hwdst=dot11_frame.getlayer(ARP).hwsrc,
            	                                    pdst=dot11_frame.getlayer(ARP).psrc)
    
                dot11_answer /= dot11_frame.getlayer(ARP).payload
            
                if options.debug:
                    os.write(1,"Sending ARP Reply on %s\n" % optios.out_iface)
                    if options.verb:
                        os.write(1,"%s\n" % dot11_answer.summary())
            
                # Frame injection :
                sendp(dot11_answer,verbose=0) # Send frame
    
    # Program killed
    except KeyboardInterrupt:
        print "Stopped by user."
    
    s.close()
    
    sys.exit()


wifidns.py
#! /usr/bin/env python

########################################
#
# wifidns.py --- WiFi injection DNS answering tool based on Wifitap
#
# Copyright (C) 2005 Cedric Blancher <sid@rstack.org>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation; version 2.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
#########################################

import os
import sys
import argparse
import struct
import logging
import utils

from socket import *
from fcntl  import ioctl
from select import select

logging.getLogger("scapy").setLevel(1)

#from scapy  import Raw,Ether,PrismHeader,Dot11,Dot11WEP,LLC,SNAP,sendp,conf
from scapy.all import Raw,Ether,PrismHeader,Dot11,Dot11WEP,LLC,SNAP,sendp,conf,RadioTap
from scapy.all  import IP,UDP,DNS,DNSRR

def setup():

    global conf

    parser = argparse.ArgumentParser()

    parser.add_argument('-b',
                    dest='bssid',
                    type=utils.sanitize_bssid,
                    required=True,
                    help='Specify BSSID for injection')

    parser.add_argument('-i',
                    dest='in_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for injection (default: wlan0)')

    parser.add_argument('-o',
                    dest='out_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for injection (default: wlan0)')

    parser.add_argument('-a',
                    dest='ipdns',
                    required=True,
                    type=utils.sanitize_ipdns,
                    help="Specify IP address for DNS answers.")

    parser.add_argument('-s',
                    dest='smac',
                    type=utils.sanitize_smac,
                    help='Specify source MAC address for injected frames.')

    parser.add_argument('-w',
                    dest='wepkey',
                    type=str,
                    help='WEP mode and key')

    parser.add_argument('-k',
                    dest='keyid',
                    type=int,
                    choices=xrange(0,4),
                    default=0,
                    help='WEP key id (default: 0)')

    parser.add_argument('-d',
                    dest='debug',
                    action='store_true',
                    help='Activate debug mode.')

    parser.add_argument('-v',
                    dest='verbe',
                    action='store_true',
                    help='Use verbose debugging.')

    parser.add_argument('-t',
                    dest='ttl',
                    type=int,
                    default=64,
                    help='Set TTL (default: 64)')

    args = parser.parse_args()

    # validate wepkey if WEP in use
    if args.wepkey is not None:
        wepkey = utils.parse_wep_key(args.wepkey, args.keyid)
        if wepkey is None:
            parser.print_usage()
            print
            print '[!] Invalid WEP key'
            sys.exit()
    
    conf.iface = args.out_iface
    conf.wepkey = args.wepkey

    return args

def print_options(options):

    print "out_iface:  %s" % options.out_iface
    print "bssid:      %s" % options.bssid
    print "ttl:        %s" % options.ttl
    print "ipdns:      %s" % options.ipdns
    if options.smac is not None:
    
        print "smac:       %s" % options.smac
    
    if options.wepkey is not None:

        print "WEP key:    %s (%dbits)" % (options.wepkey, len(options.wepkey)*4)

    print 'DEBUG mode:', options.debug

if __name__ == '__main__':

    options = setup()

    # Here we put a BPF filter so only 802.11 Data/to-DS frames are captured
    s = conf.L2listen(iface=options.in_iface,
                    filter="link[0]&0xc == 8 and link[1]&0xf == 1")
    
    # Speed optimization si Scapy does not have to parse payloads
    DNS.payload_guess=[]
    
    try:
        while True:
            dot11_frame = s.recv(2346)
            
            # WEP handling is automagicly done by Scapy if conf.wepkey is set
            # Nothing to do to decrypt
            # WEP frames have Dot11WEP layer, others don't
            if options.debug and options.verb:
                if dot11_frame.haslayer(Dot11WEP): # WEP frame
                    os.write(1,"Received WEP from %s\n" % options.in_iface)
                else: # Cleartext frame
                    os.write(1,"Received from %s\n" % options.in_iface)
                    os.write(1,"%s\n" % dot11_frame.summary())
            
            if dot11_frame.getlayer(Dot11).addr1 != options.bssid:
                continue
            
            # Identifying DNS Queries
            if dot11_frame.haslayer(DNS) and dot11_frame.getlayer(DNS).qr == 0:
                if options.debug:
                    os.write(1,"Received DNS Query on %s\n" % options.in_iface)
                    if options.verb:
                        os.write(1,"%s\n" % dot11_frame.summary())
            
                # Building DNS Reply answer for injection
                dot11_answer = RadioTap()/Dot11(type="Data",
                                            FCfield="from-DS",
                                            addr1=dot11_frame.getlayer(Dot11).addr2,
                                            addr2=options.bssid)
    
                if options.smac is not None:
                    dot11_answer.addr3 = dot11_frame.getlayer(Dot11).addr1
                else:
                    dot11_answer.addr3 = options.smac
    
                if options.wepkey is not None:
                    dot11_answer.FCfield |= 0x40
                    dot11_answer /= Dot11WEP(iv="111", keyid=options.keyid)
    
                dot11_answer /= LLC(ctrl = 3)/SNAP()/IP(src=dot11_frame.getlayer(IP).dst,
                                                    dst=dot11_frame.getlayer(IP).src,
                                                    ttl=options.ttl)
    
                dot11_answer /= UDP(sport=dot11_frame.getlayer(UDP).dport,
                                    dport=dot11_frame.getlayer(UDP).sport)
    
                dot11_answer /= DNS(id=dot11_frame.getlayer(DNS).id,
                                    qr=1,
                                    qd=dot11_frame.getlayer(DNS).qd,
                                    an=DNSRR(rrname=dot11_frame.getlayer(DNS).qd.qname,
                                        ttl=10,
                                        rdata=options.ipdns))
                 
                if options.debug:
                    os.write(1,"Sending DNS Reply on %s\n" % options.out_iface)
                    if options.verb:
                        os.write(1,"%s\n" % dot11_frame.summary())
                 
                # Frame injection :
                sendp(dot11_answer,verbose=0) # Send frame
    
    # Program killed
    except KeyboardInterrupt:
        print "\nStopped by user."
    
    s.close()
    sys.exit()


wifitap.py
#! /usr/bin/env python

########################################
#
# wifitap.py --- WiFi injection tool through tun/tap device
#
# Copyright (C) 2005 Cedric Blancher <sid@rstack.org>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation; version 2.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
#########################################

import os
import sys
import struct
import logging
import argparse
import utils

from socket import *
from fcntl  import ioctl
from select import select

logging.getLogger("scapy").setLevel(1)
#from scapy  import Raw,Ether,PrismHeader,Dot11,Dot11WEP,LLC,SNAP,sendp,conf
from scapy.all import Raw,Ether,PrismHeader,Dot11,Dot11WEP,LLC,SNAP,sendp,conf,RadioTap

TUNSETIFF = 0x400454ca
IFF_TAP   = 0x0002
TUNMODE   = IFF_TAP

def setup():

    global conf

    parser = argparse.ArgumentParser()
    parser.add_argument('-b',
                    dest='bssid',
                    type=utils.sanitize_bssid,
                    required=True,
                    help='Specify BSSID for injection')

    parser.add_argument('-o',
                    dest='out_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for injection (default: wlan0)')

    parser.add_argument('-i',
                    dest='in_iface',
                    default='wlan0',
                    type=str,
                    help='Specify iface for listening (default: wlan0)')

    parser.add_argument('-s',
                    dest='smac',
                    type=utils.sanitize_smac,
                    help='Specify source MAC address for injected frames.')

    parser.add_argument('-w',
                    dest='wepkey',
                    type=str,
                    help='WEP mode and key')

    parser.add_argument('-k',
                    dest='keyid',
                    type=int,
                    choices=xrange(0,4),
                    default=0,
                    help='WEP key id (default: 0)')

    parser.add_argument('-d',
                    dest='debug',
                    action='store_true',
                    help='Activate debug mode.')

    parser.add_argument('-v',
                    dest='verb',
                    action='store_true',
                    help='Use verbose debugging.')

    args = parser.parse_args()

    # validate wepkey if WEP in use
    if args.wepkey is not None:
        wepkey = utils.parse_wep_key(args.wepkey, args.keyid)
        if wepkey is None:
            parser.print_usage()
            print
            print '[!] Invalid WEP key'
            sys.exit()
    
    conf.iface = args.out_iface
    conf.wepkey = args.wepkey

    return args

def print_options(options):

    print "in_iface:   %s" % options.in_iface
    print "out_iface:  %s" % options.out_iface
    print "bssid:      %s" % options.bssid
    if options.smac is not None:
    
        print "smac:       %s" % options.smac
    
    if options.wepkey is not None:

        print "WEP key:    %s (%dbits)" % (options.wepkey, len(options.wepkey)*4)

    print 'DEBUG mode:', options.debug

if __name__ == '__main__':

    options = setup()
    print_options(options)
    
    # Here we put a BPF filter so only 802.11 Data/to-DS frames are captured
    s = conf.L2listen(iface = options.in_iface,
                    filter = "link[0]&0xc == 8 and link[1]&0xf == 1")
    #s = conf.L2listen(iface = options.in_iface)
    
    # Open /dev/net/tun in TAP (ether) mode
    f = os.open("/dev/net/tun", os.O_RDWR)
    ifs = ioctl(f, TUNSETIFF, struct.pack("16sH", "wj%d", TUNMODE))
    ifname = ifs[:16].strip("\x00")
    print "Interface %s created. Configure it and use it" % ifname
    
    # Speed optimization si Scapy does not have to parse payloads
    Ether.payload_guess=[]
    SNAP.payload_guess=[]
    
    try:
    
        while True:
    
            r = select([f,s],[],[])[0]
        
            # frame from /dev/net/tun
            if f in r:
            
                # tuntap frame max. size is 1522 (ethernet, see RFC3580) + 4
                buf = os.read(f,1526)
                eth_rcvd_frame=Ether(buf[4:])
                
                if options.debug:
                    os.write(1,"Received from %s\n" % ifname)
                    if options.verb:
                        os.write(1,"%s\n" % eth_rcvd_frame.summary())
                
                # Prepare Dot11 frame for injection
                dot11_sent_frame = RadioTap()/Dot11(
                type = "Data",
                FCfield = "from-DS",
                addr1 = eth_rcvd_frame.getlayer(Ether).dst,
                addr2 = options.bssid)
                # It doesn't seem possible to set tuntap interface MAC address
                # when we create it, so we set source MAC here
                if options.smac is None:
                    dot11_sent_frame.addr3 = eth_rcvd_frame.getlayer(Ether).src
                else:
                    dot11_sent_frame.addr3 = options.smac
                if options.wepkey is not None:
                    dot11_sent_frame.FCfield |= 0x40
                    dot11_sent_frame /= Dot11WEP(iv = "111", keyid = options.keyid)
    
                dot11_sent_frame /= LLC(ctrl = 3)/SNAP(code=eth_rcvd_frame.getlayer(Ether).type)/eth_rcvd_frame.getlayer(Ether).payload
                
                if options.debug:
                    os.write(1,"Sending from-DS to %s\n" % options.out_iface)
                    if options.verb:
                        os.write(1,"%s\n" % dot11_sent_frame.summary())
                
                # Frame injection :
                sendp(dot11_sent_frame,verbose=0) # Send from-DS frame
            
            # Frame from WiFi network
            if s in r:
            
                # 802.11 maximum frame size is 2346 bytes (cf. RFC3580)
                # However, WiFi interfaces are always MTUed to 1500
                dot11_rcvd_frame = s.recv(2346)
            
                # WEP handling is automagicly done by Scapy if conf.wepkey is set
                # Nothing to do to decrypt (although not yet tested)
                # WEP frames have Dot11WEP layer, others don't
            
                if options.debug:
                    if dot11_rcvd_frame.haslayer(Dot11WEP): # WEP frame
                        os.write(1,"Received WEP from %s\n" % options.in_iface)
                    else: # Cleartext frame
                        os.write(1,"Received from %s\n" % options.in_iface)
                    if options.verb:
                        os.write(1,"%s\n" % dot11_rcvd_frame.summary())
            
                # if dot11_frame.getlayer(Dot11).FCfield & 1: # Frame is to-DS
                # For now, we only take care of to-DS frames...
                if dot11_rcvd_frame.getlayer(Dot11).addr1 != options.bssid:
                    if options.verb:
                        os.write(1,"Frame not to/from BSSID\n")
                    continue
            
                # One day, we'll try to take care of AP to DS trafic (cf. TODO)
                #    else: # Frame is from-DS
                #        if dot11_frame.getlayer(Dot11).addr2 != options.bssid:
                #            continue
                #   eth_frame = Ether(dst=dot11_frame.getlayer(Dot11).addr1,
                #           src=dot11_frame.getlayer(Dot11).addr3)
                
                if dot11_rcvd_frame.haslayer(SNAP):
                    eth_sent_frame = Ether(dst=dot11_rcvd_frame.getlayer(Dot11).addr3,
                                        src=dot11_rcvd_frame.getlayer(Dot11).addr2,
                                        type=dot11_rcvd_frame.getlayer(SNAP).code)
                    eth_sent_frame.payload = dot11_rcvd_frame.getlayer(SNAP).payload
            
                if options.debug:
                    os.write(1, "Sending to %s\n" % ifname)
                    if options.verb:
                        os.write(1, "%s\n" % eth_sent_frame.summary())
            
                # Add Tun/Tap header to frame, convert to string and send
                buf = "\x00\x00" + struct.pack("!H",eth_sent_frame.type) + str(eth_sent_frame)
                os.write(f, buf)
    
    # Program killed
    except KeyboardInterrupt:
        print "\nStopped by user."
    
    s.close()
    os.close(f)
    
    sys.exit()
